src/core/toPath.js

```diff
@@ -102,139 +110,84 @@ const allowingArrays = fn => arg => {
-        if (str.charAt(index) === '.') index++
-
-        // Shorthand: if array index is the whole slice add it to path
-        if (arrayIndexValue === ':') {
-          path.push([undefined, undefined])
-        } else {
-
-          // Look for a slice quote
-          const sliceDelimIndex = arrayIndexValue.indexOf(':')
-
-          // If no slice quote found
-          if (sliceDelimIndex === -1) {
-            // Parse array index as a number
-            const nArrayIndexValue = Number(arrayIndexValue)
-
-            // Add array index to path, either as a valid index (positive int), or as a string
-            path.push(isIndex(nArrayIndexValue) ? nArrayIndexValue : arrayIndexValue)
-
-          } else { // If a slice quote is found
-
-            // Fetch slice start and end, and parse them as slice indexes (empty or valid int)
-            const sliceStart = arrayIndexValue.substring(0, sliceDelimIndex), sliceEnd = arrayIndexValue.substring(sliceDelimIndex + 1)
-            const nSliceStart = toSliceIndex(sliceStart), nSliceEnd = toSliceIndex(sliceEnd)
-
-            // Add array index to path, as a slice if both slice indexes are valid (undefined or int), or as a string
-            path.push(isSliceIndex(nSliceStart) && isSliceIndex(nSliceEnd) ? [nSliceStart, nSliceEnd] : arrayIndexValue)
-          }
-        }
+  const path = stringToPath2(str)
+  return str[0] === '.' ? ['', ...path] : path
+}
 
-        // Stop if end of string has been reached
-        if (index === str.length) break
-      }
-    }
+/**
+ * @typedef {function(string): string[]} Matcher a function that can replace String.prototype.match
+ * @param {string} str string to match against
+ * @param {[(Matcher | RegExp), function(string[]): *]} matchers
+ *   pairs of a regexp to match str against, and a function to transform the resulting match object into the final result
+ * @param {*} defaultResult
+ *   value to return if no matcher matches
+ * @returns {*} output value of the first cond that matches or defaultResult if no cond matches
+ */
+const match = (str, matchers, defaultResult) => {
+  for (const [matcher, mapper] of matchers) {
+    const match = matcher instanceof RegExp ? str.match(matcher) : matcher(str)
+    if (match) return mapper(match.slice(1))
+  }
+  return defaultResult
+}
 
+match.andCheck = (matcher, predicate) => {
+  return str => {
+    const match = str.match(sliceNotation)
+    if (!match) return match
+    return predicate(match.slice(1)) ? match : null
   }
+}
 
-  return path
+const quotedBracketNotation = /^\[(['"])(.*?[^\\])\1\]?\.?(.*)$/
+const incompleteQuotedBracketNotation = /^\[["'](.*)$/
+const sliceNotation = /^\[([^:\]]*):([^:\]]*)\]\.?(.*)$/
+const bareBracketNotation = /^\[([^\]]*)\]\.?(.*)$/
+const incompleteBareBracketNotation = /^\[(.*)$/
+const pathSegmentEndedByDot = /^([^.[]*?)\.(.*)$/
+const pathSegmentEndedByBracket = /^([^.[]*?)(\[.*)$/
+
+const stringToPath2 = str => {
```

Comment:
What would be a better name for this?

Reply:
applyMatchers ?

Reply:
Well I'd say that's more of a "how" than "what". This function actually turns a path string into a path array. The difference with `stringToPath` is that it won't prepend the path array with an empty string if the path string starts with a dot.

Reply:
Yes but the second one can have a technical name link to his context and implementation in this particular file because it doesn't have vocation to be used alone. So only the first one need to have a name that reflect the global purpose.

Reply:
I guess you're right it doesn't matter that much. I'd still prefer something more honest. Maybe `stringToPath` should be `parsePath` and `stringToPath2` should be `parsePathIgnoringLeadingDot`. Or maybe `stringToPath2` should be `parsePath` and `stringToPath` should a generic decorator instead of an explicit caller of `parsePath`. Like `allowingArrays` is. By the way the memoizing part should be a decorator too. The result would be something like:

```js
const toPath = allowingArrays(...memoize(withMeaningfulLeadingDot(parsePath)))
```

With `withMeaningfulLeadingDot` being the current `stringToPath`.

Reply:
Hey ! I just added a comment in [#113](/Zenika/immutadot/issues/113#issuecomment-343871606), I think the leading dot should just be discarded, what do you think ?

Reply:
I'm fine with it.

Reply:
Done